### PR TITLE
Set active device using respective backend native id's

### LINKED ISSUE #780: Add unit tests for statistics wrapper functions
* [x] mean
* [x] var
* [x] stdev
* [x] cov
* [x] median
* [x] corrcoef

### LINKED ISSUE #980: Add the ability to set devices using native Ids for CUDA and OpenCL
For OpenCL:

```
af_err afcl_set_device_id(cl_device_id id);
void setDeviceId(cl_device_id id);
```

For CUDA:

```
af_err afcu_set_native_id(int id);
void setNativeId(int id);
```

### PR DESCRIPTION
**Pull request by 9prady9**
Fixes #980 
Fixes #780

### CODE REVIEW

**Comment by 9prady9** [comment]
I have forked this branch from `stats_tests`, hence the huge number of commits, these will go away as soon as #1033 is merged. Please merge this after #1033

**Comment by shehzan10** [comment]
I'm assuming its just the last 2 commits. Why dont you push them to #1033 or if #1033 is included here, then close #1033 and we'll merge this.

File: src/api/c/stats.h [hunk 1]
```diff
@@ -40,39 +40,52 @@ struct baseOutType {
                                 float>::type type;
 };
 
-template<typename T>
-inline T mean(const Array<T>& in)
+template<typename Ti, typename To>
+inline To mean(const Array<Ti>& in)
```
Comment (shehzan10) [inline src/api/c/stats.h]: I think you will have to make these statics. That's why windows is failing I believe.

Comment (9prady9) [inline src/api/c/stats.h]: Not these, the other mean which has specilizations. Just did that. Also fixed the covariance unit test that was failing on OSX.

**Comment by 9prady9** [comment]
is `fft2_real` supposed to fail on OSX ?

**Comment by 9prady9** [comment]
Finally, @shehzan10 This is ready for merge.

**Pull request merged**